### PR TITLE
Balder/compile euclidian distance for avx2 and avx512

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst PR

### CODE REVIEW

**Review by geirst — COMMENTED** [review]
Overall looks good, some comments to consider. Note that "Euclidian" should be renamed to "Euclidean".

File: searchlib/src/vespa/searchlib/tensor/distance_functions.h [hunk 1]
```diff
@@ -13,66 +14,20 @@ namespace search::tensor {
 template <typename FloatType>
 class SquaredEuclideanDistance : public DistanceFunction {
 public:
+    SquaredEuclideanDistance()
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/tensor/distance_functions.h]: Fixed

File: vespalib/src/tests/hwaccelrated/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/tests/hwaccelrated/CMakeLists.txt]: Fixed

File: vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp]: Fixed

File: vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+std::vector<T> createAndFill(size_t sz) {
+    std::vector<T> v(sz);
+    for (size_t i(0); i < sz; i++) {
+        v[i] = i;
+    }
+    return v;
+}
+
+template<typename T>
+void verifyEuclidianDistance(const hwaccelrated::IAccelrated & accel) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp]: Fixed

File: vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        T sum(0);
+        for (size_t i(j); i < testLength; i++) {
+            sum += (a[i] - b[i]) * (a[i] - b[i]);
+        }
+        T hwComputedSum(accel.squaredEuclidianDistance(&a[j], &b[j], testLength - j));
+        EXPECT_EQUAL (sum, hwComputedSum);
+    }
+}
+
+TEST("test euclidian distance") {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp]: Fixed

File: vespalib/src/vespa/vespalib/hwaccelrated/avx2.cpp [hunk 6]
```diff
@@ -10,4 +10,14 @@ Avx2Accelrator::populationCount(const uint64_t *a, size_t sz) const {
     return helper::populationCount(a, sz);
 }
 
+double
+Avx2Accelrator::squaredEuclidianDistance(const float * a, const float * b, size_t sz) const {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/hwaccelrated/avx2.cpp]: Fixed

File: vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        v[i] = i;
+    }
+    return v;
+}
+
+template<typename T>
+void verifyEuclidianDistance(const hwaccelrated::IAccelrated & accel) {
+    const size_t testLength(255);
+    std::vector<T> a = createAndFill<T>(testLength);
+    std::vector<T> b = createAndFill<T>(testLength);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp]: Random with fixed seed.

File: vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp [hunk 8]
```diff
@@ -65,8 +62,24 @@ void verifyAccelrator(const IAccelrated & accel)
...
     }
-    delete [] a;
-    delete [] b;
+}
+
+template<typename T>
+void verifyEuclidianDistance(const IAccelrated & accel) {
+    const size_t testLength(255);
+    std::vector<T> a = createAndFill<T>(testLength);
+    std::vector<T> b = createAndFill<T>(testLength);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/hwaccelrated/iaccelrated.cpp]: Random with fixed seed.

**Review by geirst — COMMENTED** [review]
👍

**Pull request merged**